src/Tests/Microsoft.Diagnostics.Monitoring.HostingStartup.UnitTests/ParameterCapturing/SystemLoggingTests.cs

```diff
@@ -29,6 +29,8 @@ public void TestLoggingCategories()
             {
                 logger.Log(ParameterCaptureMode.Inline, inlineMessage, Array.Empty<string>());
                 logger.Log(ParameterCaptureMode.Background, backgroundMessage, Array.Empty<string>());
+
+                logger.Complete();
```

Comment:
I don't understand how this currently helps with tests; the shutdown ordering and logic remains the same.

Reply:
I will add more comments in the code, but effectively `Complete()` will cause all the messages to be processed prior to the Dispose operation. Prior to this change, the dispose flag would get set immediately and potentially cancel the background thread before it finished processing.

Reply:
Okay, I see that now in `ThreadProc` it is checking the disposal state and keying off that to determine whether it should stop processing entries, which is... interesting. Not sure how I feel about that since it's somewhat conflating completion with disposal. I think it would be better if it checked for `_messages.IsAddingCompleted` or (maybe better, if we want it to finish processing all of the messages) `_messages.IsCompleted`.

Reply:
Or even use `while (_messages.TryTake(out (string format, string[] args), Timeout.InfiniteTimeSpan))` to have it wait for an item and return false when no more items are available rather than likely throwing an exception due to waiting in `_messages.Take` when it gets completed or disposed.